File: src/sustech-lambda/src/main/java/cn/edu/sustc/cse/ooad/sustechlambda/sustechlambda/security/JwtAuthenticationFilter.java
package cn.edu.sustc.cse.ooad.sustechlambda.sustechlambda.security;

import io.jsonwebtoken.Jwts;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class JwtAuthenticationFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request,
                                    HttpServletResponse response,
                                    FilterChain filterChain)
            throws ServletException, IOException {
        var token = extractToken(request);
        if (token != null) {
            var claims = Jwts.parser()
                    .setSigningKey("my super signing key".getBytes())
                    .parseClaimsJws(token)
                    .getBody();
            var authenticationToken = new JwtAuthenticationToken(claims);
            SecurityContextHolder.getContext().setAuthentication(authenticationToken);
        }
        filterChain.doFilter(request, response);
    }

    private String extractToken(HttpServletRequest request) {
        var authHeader = request.getHeader("Authorization");
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            return authHeader.substring(7);
        }
        return null;
    }
}

File: src/sustech-lambda/src/main/kotlin/cn/edu/sustc/cse/ooad/sustechlambda/security/JwtAuthenticationFilter.kt
package cn.edu.sustc.cse.ooad.sustechlambda.security

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.filter.OncePerRequestFilter
import java.security.Key
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JwtAuthenticationFilter(private val signingKey: Key) : OncePerRequestFilter() {

    override fun doFilterInternal(
            request: HttpServletRequest,
            response: HttpServletResponse,
            filterChain: FilterChain
    ) {
        try {
            extractToken(request)?.let {
                Jwts.parser()
                        .setSigningKey(this.signingKey)
                        .parseClaimsJws(it)
                        .body
            }?.let {
                JwtAuthenticationToken(it)
            }.also {
                SecurityContextHolder.getContext().authentication = it
            }
            filterChain.doFilter(request, response)
        } catch (ex: MalformedJwtException) {
            response.sendError(401, "Invalid access token")
        }

    }
}

private fun extractToken(request: HttpServletRequest) = request.getHeader("Authorization")?.let {
    when {
        it.startsWith("Bearer ") -> it.substring(7)
        else -> null
    }
}
